question 4:

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 
pokeaman

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [2]:
import statsmodels.formula.api as smf

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Mon, 11 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        01:10:36   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Goodness of fit (R²) and statistical significance of the coefficients are two separate independent aspects for evaluating model. A low R² means the model doesn't explain much of the variability, but it doesn't rule out the possibility that individual predictors have strong, significant effects on the outcome.

The Model Explains 17.6% of the Variability (R²):
1) R²: the proportion of variance in dependent variable which explained by independent variable.
2) R² of 17.6%: it shows that the model explains 17.6% of variablity in HP variablity. It is quite low, and it means that the predictors cannot explain most of the variation in HP. It doesn’t mean that the predictors have no impact.
3) R² cannot capture the effect of individual predictors for making huge contributions to the model, it gives an idea of goodness of fit

Strong Evidence Against the Null Hypothesis (Statistical Significance of Coefficients):
1) "many of the coefficients are larger than 10 while having strong or very strong evidence against the null hypothesis" refers to statistical significance. This means independent variable (Sp. Def, Generation) will have a real effect on dependent variable (HP). This is indicated by their p-value
2) Large coefficients reflect that predictors have a huge impact on the dependent variable. The lower p-value indicates that the predictors are not equal to zero (no effect).
This means while the model doesn't explain much of the variation in HP, the predictors can still have a statistically huge influence on the dependent variable.

The Possible Cause of the Contradiction:
1) Low R² (17.6%): this shwos that predictors do not capturing all the factors that affect HP in the model. The reason to causing it is: first, some important variables missing in the model. second, the relationship between predictors and outcomes are not strong.
2) Significant Coefficients (low p-values and large values): Even though there is a low R², the predictors have a strong impact on independent variables. The reason of causing it is: 1. predictors influences the outcomes, but there are many unexplained variability that the model cannot account for. 2. There are other factors (variables) influence the HP but not being included in the data.  

Key takeaways:
A low R² suggests the model does not explain a lot related to total variablity in HP, but it does indicates the predictors are not important. The statistical significant of predictors shows the relationship between predictors and outputs are real, the output is not due the random chance. 
This contradiction shows model fit (R²) and statistical significance are different. Model can have a statistically significant predictors but only explain a small amount of total variability in the output.

Summary and link:

The model you fit shows a low R-squared value (17.6%), meaning it explains only a small portion of the variability in the 𝐻𝑃 variable. However, several coefficients are large and statistically significant, which might seem contradictory. This occurs because R-squared and statistical significance measure different things:

R-squared indicates the proportion of variance in the dependent variable (HP) explained by the model, so a low R-squared suggests that Sp.Def and Generation don't capture much of the variation in HP.

Statistical Significance of Coefficients reflects the likelihood that the observed associations between predictors and HP are due to chance. Significant coefficients imply there is a non-random relationship between these predictors and HP, even if the model's overall explanatory power is low.

In essence, the predictors in your model show statistically significant associations with HP, but they do not account for much of its variability. This could happen if other factors not included in the model play a larger role in explaining HP.

https://chatgpt.com/share/673154e3-1c1c-8013-ad0c-dba6fa930c96

Question 7

In [4]:
import numpy as np

In [6]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import statsmodels.formula.api as smf

# Load the dataset
url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
pokeaman = pd.read_csv(url)

# Replace NaN values in the "Type 2" column with "None"
pokeaman.fillna('None', inplace=True)

# Split the data into training and testing sets
np.random.seed(130)
fifty_fifty_split_size = int(pokeaman.shape[0] * 0.5)
pokeaman_train, pokeaman_test = train_test_split(pokeaman, train_size=fifty_fifty_split_size)

# Define the model
model5_linear_form = 'HP ~ Attack + Defense + Speed + Legendary + Q("Sp. Def") + Q("Sp. Atk") + C(Generation) + C(Q("Type 1")) + C(Q("Type 2"))'
model5_spec = smf.ols(formula=model5_linear_form, data=pokeaman_train)
model5_fit = model5_spec.fit()

# Display the summary
print(model5_fit.summary())

yhat_model5 = model5_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model5_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model5)[0,1]**2)

                            OLS Regression Results                            
Dep. Variable:                     HP   R-squared:                       0.392
Model:                            OLS   Adj. R-squared:                  0.313
Method:                 Least Squares   F-statistic:                     4.948
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           9.48e-19
Time:                        02:23:46   Log-Likelihood:                -1765.0
No. Observations:                 400   AIC:                             3624.
Df Residuals:                     353   BIC:                             3812.
Df Model:                          46                                         
Covariance Type:            nonrobust                                         
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
Intercept           

Model 5:
Rationale: In model5_linear_form, the main prefictors are Speed, Sp. Def, Attack, Defense, and Sp. Atk, categorical indicators are Type 1, Type 2, Legendary, and Generation.
Objective: The model5_linear_form use a lot of prefictors to capture the effect on HP, then it establish initial comprehensive baseline for prediction.

In [7]:
# Here's something a little more reasonable...
model6_linear_form = 'HP ~ Attack + Speed + Q("Sp. Def") + Q("Sp. Atk")'
# And here we'll add the significant indicators from the previous model
# https://chatgpt.com/share/81ab88df-4f07-49f9-a44a-de0cfd89c67c
model6_linear_form += ' + I(Q("Type 1")=="Normal")'
model6_linear_form += ' + I(Q("Type 1")=="Water")'
model6_linear_form += ' + I(Generation==2)'
model6_linear_form += ' + I(Generation==5)'

model6_spec = smf.ols(formula=model6_linear_form, data=pokeaman_train)
model6_fit = model6_spec.fit()
model6_fit.summary()

yhat_model6 = model6_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model6_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2)

'In sample' R-squared:     0.3326310334310908
'Out of sample' R-squared: 0.29572460427079933


Model 6:
Developement: model6_linear_form focuses on the most impactful prefictors which are: Speed, Attack, Sp. Def, Sp. Atk, and includes the significant categorical indictors which are: Generations (2 and 5) and Type 1 (Normal and Water), etc.
Rationale:
The model refines predictor through eliminates less significant variables in order to reduce complexity. This improves model interpretability and prediction accuracy.

In [8]:
# And here's a slight change that seems to perhaps improve prediction...
model7_linear_form = 'HP ~ Attack * Speed * Q("Sp. Def") * Q("Sp. Atk")'
model7_linear_form += ' + I(Q("Type 1")=="Normal")'
model7_linear_form += ' + I(Q("Type 1")=="Water")'
model7_linear_form += ' + I(Generation==2)'
model7_linear_form += ' + I(Generation==5)'

model7_spec = smf.ols(formula=model7_linear_form, data=pokeaman_train)
model7_fit = model7_spec.fit()
model7_fit.summary()
yhat_model7 = model7_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model7_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2)

'In sample' R-squared:     0.37818209127432456
'Out of sample' R-squared: 0.35055389205977444


In [9]:
# And here's a slight change that seems to perhas improve prediction...
model7_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Speed))'
model7_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# We DO NOT center and scale indicator variables
model7_linear_form_CS += ' + I(Q("Type 1")=="Normal")'
model7_linear_form_CS += ' + I(Q("Type 1")=="Water")'
model7_linear_form_CS += ' + I(Generation==2)'
model7_linear_form_CS += ' + I(Generation==5)'

model7_CS_spec = smf.ols(formula=model7_linear_form_CS, data=pokeaman_train)
model7_CS_fit = model7_CS_spec.fit()
model7_CS_fit.summary().tables[-1] 
# "Cond. No." is NOW 15.4 due to centering and scaling

# "Cond. No." WAS 2,340,000,000 WITHOUT to centering and scaling
model7_fit.summary().tables[-1]

<class 'statsmodels.iolib.table.SimpleTable'>

Model 7:
Further Refinement and Interaction Terms: model7_linear_form introduces interaction terms between Attack, Speed, Sp. Def, and Sp. Atk alongside significant indicators.
Objective: Adding interactions allows the model to capture potential synergistic effects between continuous predictors, and this explains more variability in HP.

Model 7 with centering and scaling:
Multicollinearity Addressed: model7_linear_form_CS introduces interaction terms between Speed, Attack, Sp. Def, and Sp. Atk alongside significant indicators.
Outcome: Scaling and centering to continuous predictors improves robustness by mitigating the effects of multicollinearity. This enhances out-of-sample performance stability.

summary and links:

In this progression of linear models, each version builds upon the last to improve predictive accuracy and model stability:

Model 5 includes a wide range of variables, providing a comprehensive baseline with various attributes (e.g., Attack, Defense, Speed) and categorical factors (Generation, Type 1, Type 2).

Model 6 refines Model 5 by excluding certain variables and adding indicator terms for significant types (e.g., "Normal", "Water") and specific generations (Generation == 2 and Generation == 5), allowing a more targeted approach to capture relevant predictors without unnecessary complexity.

Model 7 introduces multiplicative interactions between key predictors (e.g., Attack * Speed * Sp. Def * Sp. Atk) to account for nonlinear relationships, enhancing model flexibility. A variant with centered and scaled terms (Model 7_CS) reduces multicollinearity, resulting in a lower condition number and improved stability.

This progression results in a streamlined model that balances complexity, predictive strength, and stability by refining variable selection, interactions, and scaling techniques.

https://chatgpt.com/share/6736b6c9-a320-8013-ba40-d701de8284fe

Question 9

In [10]:
model7_gen1_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model7_gen1_predict_future_fit = model7_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model7_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

'In sample' R-squared:     0.37818209127432456 (original)
'Out of sample' R-squared: 0.35055389205977444 (original)
'In sample' R-squared:     0.5726118179916575 (gen1_predict_future)
'Out of sample' R-squared: 0.11151363354803218 (gen1_predict_future)


In [11]:
model7_gen1to5_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model7_gen1to5_predict_future_fit = model7_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model7_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

'In sample' R-squared:     0.37818209127432456 (original)
'Out of sample' R-squared: 0.35055389205977444 (original)
'In sample' R-squared:     0.3904756578094535 (gen1to5_predict_future)
'Out of sample' R-squared: 0.23394915464343125 (gen1to5_predict_future)


In [12]:
model6_gen1_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model6_gen1_predict_future_fit = model6_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model6_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

'In sample' R-squared:     0.3326310334310908 (original)
'Out of sample' R-squared: 0.29572460427079933 (original)
'In sample' R-squared:     0.4433880517727282 (gen1_predict_future)
'Out of sample' R-squared: 0.1932858534276128 (gen1_predict_future)


In [13]:
model6_gen1to5_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model6_gen1to5_predict_future_fit = model6_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model6_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

'In sample' R-squared:     0.3326310334310908 (original)
'Out of sample' R-squared: 0.29572460427079933 (original)
'In sample' R-squared:     0.33517279824114776 (gen1to5_predict_future)
'Out of sample' R-squared: 0.26262690178799936 (gen1to5_predict_future)


This code demonstrates how well a machine learning model can generalize across different subsets of data. By training the model on data from specific Pokémon generations (such as Generation 1) and testing it on other generations (such as Generation 6), we can observe how well it adapts to unseen data. The comparison of "in-sample" (training data) and "out-of-sample" (test data) R-squared values helps us understand if the model is overfitting, meaning it may perform well on the training data but poorly on new data.

summary and link:

The code you've shared involves fitting linear regression models using OLS (Ordinary Least Squares) to predict the HP (hit points) of Pokémon based on their generation. Specifically, the models are evaluated in two key ways:

'In Sample' R-Squared: Measures how well each model fits the data it was trained on (the training set). This gives an indication of how well the model explains the variance in the data it was provided.

'Out of Sample' R-Squared: Evaluates how well the model generalizes to new, unseen data (the test set). This is calculated by comparing the predicted values from the model with the actual values from the test set, and it indicates the model's ability to predict correctly for new data.

The models are applied to Pokémon data from different generations, specifically Generation 1 and Generations 1-5, and the performance is assessed for both in-sample and out-of-sample predictions across different subsets of the data. The comparison of the R-squared values helps to assess how well each model performs across different Pokémon generations and how it generalizes to unseen data.

By comparing models (model7 and model6), the approach seeks to understand the effectiveness of predictions for different generations of Pokémon, ensuring that the models are robust and predictive across a range of Pokémon data.

https://chatgpt.com/share/6736bb10-2f60-8013-a995-c45c4bdbdd9c